### PR TITLE
kconfig: cpp: Have LIB_CPLUSPLUS depend on NEWLIB_LIBC

### PR DESCRIPTION
**Pull request by SebastianBoe**
The C++ STD library is not compatible with the minimal C library. To
ensure that users do not accidentally enable LIB_CPLUSPLUS while also
enabling a minimal libc library we add a dependency in Kconfig.

Also, use depends instead of select between EXCEPTIONS, RTTI, and
LIB_CPLUSPLUS as use of select is discouraged in this situation.

### CODE REVIEW

**Review by galak — APPROVED** [review]

File: subsys/cpp/Kconfig
Comment (ulfalizer) [inline subsys/cpp/Kconfig]: The organization here feels a bit weird. Maybe we could do this instead:

```
config LIB_CPLUSPLUS
	bool "Link with STD C++ library"
	depends on NEWLIB_LIBC
	help
	  Link with STD C++ Library.

if LIB_CPLUSPLUS

config EXCEPTIONS
	bool "Enable C++ exceptions support"
	help
	  This option enables support of C++ exceptions.

config RTTI
	bool "Enable C++ RTTI support"
	help
	  This option enables support of C++ RTTI.

endif # LIB_CPLUSPLUS
```

Might need to fix up some configuration files too, if they enable `EXCEPTIONS` without enabling `LIB_CPLUSPLUS`.

**Review by ulfalizer — APPROVED** [review]
Fine otherwise.

Guessing the thinking is that you might be able to have exceptions or RTTI without the standard library at some point.

**Comment by ioannisg** [comment]
@galak can this be consider a bug-fix that needs to go to 2.0?

**Comment by SebastianBoe** [comment]
@ioannisg : IMHO this should not be applied in the stabilization period (because it makes things more strict and therefore can destabilize).

File: subsys/cpp/Kconfig
Comment (SebastianBoe) [reply subsys/cpp/Kconfig]: Agreed, done.

**Comment by ioannisg** [comment]
> 
> 
> @ioannisg : IMHO this should not be applied in the stabilization period (because it makes things more strict and therefore can destabilize).

I am ok with postponing this to 2.1, as long as there is no bug-report related to the fix (then we should merge that now)

**Review by aescolar — DISMISSED** [review]
Could you please rebase on top of https://github.com/zephyrproject-rtos/zephyr/pull/18315 and if needed make these options depend on !MINIMAL_LIBC?
I think that would be a much nicer solution

**Comment by SebastianBoe** [comment]
Sure, assuming #18315 goes in before this PR.

**Comment by SebastianBoe** [comment]
Blocked by #18315 . Will rebase this after 18315 is merged.

**Comment by SebastianBoe** [comment]
Rebased

**Review by pabigot — APPROVED** [review]

**Review by vanwinkeljan — APPROVED** [review]

**Comment by galak** [comment]
Needs rebasing.

**Comment by SebastianBoe** [comment]
Rebased

**Comment by SebastianBoe** [comment]
@galak : This is ready to merge

**Pull request merged**
